File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /**
     * Time Complexity: O(max(N, M)), where N and M are the lengths of l1 and l2.
     * We iterate through both lists exactly once.
     *
     * Space Complexity: O(max(N, M)), as we create a new linked list to store the sum.
     */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // 1. Initialize a dummy node. This is the simplest way to handle the head of the new list.
        ListNode dummyHead = new ListNode(0);
        
        // 'current' pointer to build the new list.
        ListNode current = dummyHead; 
        
        // 'carry' tracks the carry-over from the previous digit addition.
        int carry = 0; 

        // 2. Loop until both lists are exhausted AND there is no carry.
        while (l1 != null || l2 != null || carry != 0) {
            
            // Get digit from l1 (or 0 if l1 is null).
            int x = (l1 != null) ? l1.val : 0;
            
            // Get digit from l2 (or 0 if l2 is null).
            int y = (l2 != null) ? l2.val : 0;

            // 3. Perform the addition (current digits + carry).
            int sum = x + y + carry;

            // Calculate the new carry for the next iteration.
            carry = sum / 10;
            
            // Calculate the digit for the current node.
            int digit = sum % 10;

            // 4. Create a new node and append it to the result list.
            current.next = new ListNode(digit);
            
            // Move the current pointer forward.
            current = current.next;

            // 5. Advance the list pointers.
            if (l1 != null) {
                l1 = l1.next;
            }
            if (l2 != null) {
                l2 = l2.next;
            }
        }

        // 6. Return the actual head of the list (the node after the dummy head).
        return dummyHead.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list (provided by LeetCode)
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode(0)
        tail = dummyHead
        carry = 0

        while l1 is not None or l2 is not None or carry != 0:
            digit1 = l1.val if l1 is not None else 0
            digit2 = l2.val if l2 is not None else 0

            sum = digit1 + digit2 + carry
            digit = sum % 10
            carry = sum // 10

            newNode = ListNode(digit)
            tail.next = newNode
            tail = tail.next

            l1 = l1.next if l1 is not None else None
            l2 = l2.next if l2 is not None else None

        result = dummyHead.next
        dummyHead.next = None
        return result
        
        # 'curr' is the pointer used to build the new list.
        curr = dummy
        
        # 'carry' tracks the carry-over to the next digit.
        carry = 0
        
        # Loop as long as there are nodes in either list OR there's a carry-over.
        while l1 or l2 or carry:
            
            # Get the value of the current node or 0 if the list is exhausted.
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            # 1. Sum the digits and the carry.
            total = val1 + val2 + carry
            
            # 2. Update the carry for the next iteration (e.g., 15 // 10 = 1).
            carry = total // 10
            
            # 3. Create a new node with the current digit (e.g., 15 % 10 = 5).
            # Attach it to the result list and move the 'curr' pointer.
            curr.next = ListNode(total % 10)
            curr = curr.next
            
            # 4. Advance the pointers for the next iteration.
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
                
        # The result is the next node after the dummy head.
        return dummy.next
        